File: backend/src/classes/connection.ts
import { City, cityFromRaw } from "../enums/city-enum";
import { DayOfWeek, dayNameToEnum } from "../enums/day-of-week";
import { TrainType, trainTypeFromRaw } from "../enums/train-types";


export class Connection{
    private routeId: string;
    private cityFrom: City;
    private cityTo: City;
    private departureTime: Date;
    private arrivalTime: Date;
    private dayOffset: number;
    private daysOfOperation: Set<DayOfWeek>
    // in minutes
    private duration: number
    private trainType: TrainType;
    private firstClassPrice: number;
    private secondClassPrice: number;

    constructor(routeId: string, CityFrom: string, CityTo: string, DepartureTime: string, ArrivalTime: string, firstClassPrice: number, secondClassPrice: number, trainType: string, daysofOperation:string){
        this.routeId = routeId;
        this.cityFrom = this.parseCity(CityFrom);
        this.cityTo = this.parseCity(CityTo);
        this.departureTime = new Date(DepartureTime);
        this.parseAndSetArrivalTime(ArrivalTime);

        const durationMs = this.arrivalTime.getTime() - this.departureTime.getTime();
        const durationMinutes = durationMs / (1000 * 60);
        this.duration = durationMinutes

        this.trainType = trainTypeFromRaw[trainType]!;

        this.firstClassPrice = firstClassPrice;
        this.secondClassPrice = secondClassPrice;

        this.daysOfOperation = this.parseDaysOfOperation(daysofOperation);
    }

    private parseDaysOfOperation(daysStr: string): Set<DayOfWeek> {
        const daysSet = new Set<DayOfWeek>();

        if(daysStr.toLowerCase() === "daily") {
            Object.values(DayOfWeek).forEach(day => daysSet.add(day as DayOfWeek));
        } else if(daysStr.includes(",")) {
            const daysArray = daysStr.split(",").map(day => day.trim());
            for (const day of daysArray) {
                const dayEnum = (DayOfWeek as any)[day.toUpperCase()];
                if (dayEnum) {
                    daysSet.add(dayEnum);
                } else {
                    console.error(`Invalid day of operation: ${day}`);
                }
            }
        }else if(daysStr.includes("-")){
            const dayArr = daysStr.split("-").map(day => day.trim());
            
            if(dayArr.length !== 2){
                console.error(`Invalid range of days: ${daysStr}`);
                return daysSet;
            }

            const startDay = dayNameToEnum[dayArr[0]];
            const endDay =  dayNameToEnum[dayArr[1]];

            for(let i = startDay; ; i = (i + 1) % 7){
                daysSet.add(i);
                if(i === endDay) break;
            }
        }

        return daysSet
    }

    private parseAndSetArrivalTime(arrivalTimeStr: string): void {
        const parts = arrivalTimeStr.trim().split(" ");

        if (parts.length > 1) {
            const timePart = parts[0];
            const offsetPart = parts[1];
            const match = offsetPart.match(/\+(\d+)d/);

            if (!match || !match[1]) {
                console.error(`Invalid arrival time format: ${arrivalTimeStr}`);
                return;
            }

            const dayOffset = parseInt(match[1], 10);
            this.dayOffset = dayOffset;

            // Parse arrival time hours and minutes
            const [hours, minutes] = timePart.split(":").map(Number);

            // Create new Date based on departureTime with added days and set time
            const baseDate = new Date(this.departureTime.getTime()); // clone departure time
            baseDate.setDate(baseDate.getDate() + dayOffset);
            baseDate.setHours(hours, minutes, 0, 0);

            this.arrivalTime = baseDate;
        } else {
            // No offset, just parse time relative to departure date with dayOffset 0
            const [hours, minutes] = parts[0].split(":").map(Number);
            this.dayOffset = 0;

            const baseDate = new Date(this.departureTime.getTime());
            baseDate.setHours(hours, minutes, 0, 0);
            this.arrivalTime = baseDate;
        }
    }

    private parseCity(City: string): City {
        const cityValue = cityFromRaw[City.toLowerCase()]!;
        if (!cityValue) {
            console.error(`Invalid city value: ${City}`);
        }
        return cityValue;
    }
    
}

File: backend/transit/models/Connection.py


from transit.constants import City, city_from_raw, DayOfWeek, day_name_to_enum, train_type_from_raw as train
from datetime import datetime, timedelta
from typing import Set 

# This class basically represents a single row in the CSV file
# It holds all the information about a single train connection
# It also parses and converts the raw data into more useful formats


class Connection:
    dummy_date = datetime(2000, 1, 1)
    
    def __init__(
        self,
        route_id: str,
        departure_city: str,
        arrival_city: str,
        departure_time: str,
        arrival_time: str,
        days_of_operation: str,
        train_type: str,
        first_class_price: str,
        second_class_price: str
        ):
        self.route_id = route_id
        self.departure_city = city_from_raw[departure_city.lower()]
        self.arrival_city = city_from_raw[arrival_city.lower()]
        hour, minute = map(int, departure_time.split(":"))
        self.departure_time = Connection.dummy_date.replace(hour=hour, minute=minute)
        self.parse_set_arrival_time(arrival_time)
        self.duration = self.arrival_time - self.departure_time
        self.train_type = train[train_type]
        self.first_class_price = float(first_class_price)
        self.second_class_price = float(second_class_price)
        self.days_of_operation = self.parse_days_of_operation(days_of_operation)
        
      
    
    def parse_set_arrival_time(self, arrival_time: str):
        parts = arrival_time.split(" ")
        
        if len(parts) > 1:
              # Arrival time includes day offset, e.g., "23:45 (+1d)"
            time = parts[0]
            day_offset_str = parts[1]
            
            import re
            m = re.match(r"\(\+(\d+)d\)", day_offset_str)
            
            if not m:
                raise ValueError(f"Invalid day offset format: {day_offset_str}")
            
            self.day_offset = int(m.group(1))    
            hour, minute = map(int, time.split(":"))
        else:
            # No day offset, arrival on same dummy date
            self.day_offset = 0
            hour, minute = map(int, arrival_time.split(":"))
            
            
        
        self.arrival_time = (Connection.dummy_date + timedelta(days=self.day_offset)).replace(hour=hour, minute=minute)
    
    def parse_days_of_operation(self, days_of_operation: str) -> Set[DayOfWeek]:
        days = set()
        
        if days_of_operation == "Daily":
            # add all days of the week
            for day in DayOfWeek:
                days.add(day)
        elif( "-" in days_of_operation):
            # add the range of days
            start_str, end_str = days_of_operation.split("-")
            start_day = day_name_to_enum[start_str].value
            end_day = day_name_to_enum[end_str].value
            current_day = start_day
            
            while True:
                days.add(DayOfWeek(current_day))
                if current_day == end_day:
                    break
                current_day = (current_day+1) % 7  # wrap around using modulo
        else:
            # add individual days
            for day_str in days_of_operation.split(","):
                day = day_name_to_enum[day_str.strip()]
                days.add(day)
                
        return days